src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp

```diff
@@ -93,47 +121,86 @@ void Bank::evaluate_and_allocate() {
         }
         storage_guard.unlock();
 
-        ov::parallel_for(vec.size(), [&](std::size_t idx) {
-            const auto& lt = vec[idx];
+        std::vector<AllocatedTensor> uids_to_allocated(uid_count);
+
+        // No allocation needed
+        if (device_for_alloc == "CPU") {
+            ov::parallel_for(vec.size(), [&](std::size_t idx) {
+                const auto& lt = vec[idx];
+                std::unique_lock dev_guard(device_bank.mutex);
+                auto iter_device_registered = device_bank.registered_tensors.find(lt);
+                NPUW_ASSERT(iter_device_registered != device_bank.registered_tensors.end() &&
+                            "Tensor should be registered first!");
+                auto uid = iter_device_registered->second;
+                if (device_bank.storage[uid].tensor) {
+                    // Already allocated
+                    return;
+                }
+
+                dev_guard.unlock();
+                auto transformed = lt.eval();
+
+                // Note: No need to relock here as the storage is already prepared for writing
+                device_bank.storage.at(uid).tensor = std::move(transformed);
+                const_cast<LazyTensor&>(lt).detach();
+            });
+        } else {
+            // Allocation needed
             std::unique_lock dev_guard(device_bank.mutex);
-            auto iter_device_registered = device_bank.registered_tensors.find(lt);
-            NPUW_ASSERT(iter_device_registered != device_bank.registered_tensors.end() &&
-                        "Tensor should be registered first!");
-            if (device_bank.storage[iter_device_registered->second].tensor) {
-                // Already allocated
-                return;
+            for (std::size_t idx = 0; idx < vec.size(); ++idx) {
+                const auto& lt = vec[idx];
+                auto iter_device_registered = device_bank.registered_tensors.find(lt);
+                NPUW_ASSERT(iter_device_registered != device_bank.registered_tensors.end() &&
+                            "Tensor should be registered first!");
+                auto uid = iter_device_registered->second;
+                if (device_bank.storage[uid].tensor) {
+                    // Already allocated
+                    continue;
+                }
+
+                uids_to_allocated[uid] = {true, lt.eval_meta(), ov::Tensor()};
```

Comment:
So it is really not an "allocated" tensor. It is not allocated here. Maybe it should be called `TensorToAllocate` instead?

Reply:
Also, why do you need both `true` and an empty tensor? Do you have cases when you have `false`?

Reply:
Reworked